runtime/sema/member_set.go

```diff
@@ -18,60 +18,94 @@
 
 package sema
 
-import (
-	"github.com/raviqqe/hamt"
-
-	interfaceentry "github.com/onflow/cadence/runtime/common/interface_entry"
-)
-
-// MemberSet is an immutable set of field assignments.
+// MemberSet is set of members.
 //
 type MemberSet struct {
-	set hamt.Set
+	Parent  *MemberSet
+	members map[*Member]struct{}
 }
 
 // NewMemberSet returns an empty member set.
 //
-func NewMemberSet() *MemberSet {
-	return &MemberSet{hamt.NewSet()}
-}
-
-func (ms *MemberSet) entry(member *Member) hamt.Entry {
-	return interfaceentry.InterfaceEntry{Interface: member}
+func NewMemberSet(parent *MemberSet) *MemberSet {
+	return &MemberSet{
+		members: map[*Member]struct{}{},
+		Parent:  parent,
+	}
 }
 
 // Add inserts a member into the set.
 //
 func (ms *MemberSet) Add(member *Member) {
-	entry := ms.entry(member)
-	ms.set = ms.set.Insert(entry)
+	ms.members[member] = struct{}{}
 }
 
 // Contains returns true if the given member exists in the set.
 //
 func (ms *MemberSet) Contains(member *Member) bool {
-	entry := ms.entry(member)
-	return ms.set.Include(entry)
+	_, ok := ms.members[member]
+	if ok {
+		return true
+	}
+
+	if ms.Parent != nil {
+		return ms.Parent.Contains(member)
+	}
+
+	return false
 }
 
-// Intersection returns a new set containing all members that exist in both sets.
+// ForEach calls the given function for each member.
+// It can be used to iterate over all members of the set.
 //
-func (ms *MemberSet) Intersection(b *MemberSet) *MemberSet {
-	result := hamt.NewSet()
+func (ms *MemberSet) ForEach(cb func(member *Member) error) error {
+	memberSet := ms
+
+	visited := map[*Member]bool{}
+
+	for memberSet != nil {
+
+		for member := range memberSet.members {
+			if visited[member] {
+				continue
+			}
+
+			err := cb(member)
+			if err != nil {
+				return err
+			}
 
-	_ = ms.set.ForEach(func(entry hamt.Entry) error {
-		if !b.set.Include(entry) {
+			visited[member] = true
+		}
+
+		memberSet = memberSet.Parent
+	}
+
+	return nil
+}
+
+// Intersection returns a new set containing all members that exist in this and the given set.
+//
+func (ms *MemberSet) Intersection(otherMS *MemberSet) *MemberSet {
+
+	result := NewMemberSet(nil)
+
+	_ = ms.ForEach(func(member *Member) error {
+		if !otherMS.Contains(member) {
```

Comment:
Just a minor suggestion: inverting the condition could reduce the number of return statements to  one.

Reply:
I usually try to structure code with early exits which avoids deep nesting. This also seems to be idiomatic / often seen in Go code.